Ignore broken console output in some situations.

ehuss opened the pull request:
If stdout or stderr is closed while Cargo is running, Cargo would panic in some situations (usually with EPIPE). For example, `cargo install --list | grep -q cargo-fuzz`, where `grep` will close stdout once it gets a match. This changes it so that Cargo will ignore output errors in most situations. Failure to output a regular build message still results in an error, which follows the behavior of some tools like `make`.

All output, including stdout, now goes through `Shell`.

Closes #5234

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented:
One of my main worries about this is how we're going to keep this working over time. This is a much larger PR than I would have expected it to be, and it seems like it's going to be very difficult to remember to always use not-println and not-eprintln.

I'm also a little worried about ignoring the results of println/eprintln here too. Most unix-like tools effectively exit immediately on EPIPE instead of ignoring the result and keep going. I do think it makes sense to draw inspiration from other build tools like `make` about what to do once we're building, but otherwise if we're, for example, in resolution where some statuses may be printed do we keep going in the background once stdout/stderr are closed?

I think ideally it'd be great to say "println error means we die silently" by default and otherwise we'd whitelist a phase like the build phase where we handle println errors specifically. I don't think there's a great way to say "println error means we die silently", though, without overriding the panic handler and inspecting the panic message...

ehuss commented:
> it's going to be very difficult to remember to always use not-println and not-eprintln

It probably wouldn't be too difficult to write a test using syn that scans for println/eprintln. Would that make you feel more comfortable?

> I'm also a little worried about ignoring the results of println/eprintln here too.

My thinking here is that all of these situations result in a very small amount of output, so exiting early doesn't save very much.  Status messages (like "Downloading...") are still treated as errors if they cannot be displayed, so I believe any long-running steps (like resolution) should still exit early if stderr is closed.

If you think the early exit is important, then we could probably change the print macros to return a `CargoResult`, and then at the top level check for the equivalent of EPIPE, and ignore it (exit with 0).  I'm reluctant to do that just due to the added complexity, and like I mentioned, I don't know of any circumstances where that would have a significant improvement.  However, I'm willing to add it if you think it is important.  I don't think I would want to catch panics.

The complete list of affected commands are:
* `-Z help`
* `cargo version` and `--version`
* `cargo list` and `--list`
* `cargo pkgid`
* `cargo tree`
* `-Zfeatures=compare`
* `cargo install --list`
* `cargo package --list`
* `cargo login`
* `cargo owners --list`
* `cargo search`
* `cargo vendor` (the final "here's your config")
* JSON output:
  * `cargo locate-project`
  * `cargo metadata`
  * `cargo read-manifest`
  * `cargo verify-project`
  * `-Ztimings`
  * `--unit-graph`

AFAIK, all of these emit a limited amount of data. `cargo metadata` might be the largest. I'm not sure how tools like `jq` works, if it is capable of closing the input prematurely.  But once cargo starts sending the content out, I think it only takes a few milliseconds to finish.

alexcrichton commented:
We discussed this in the Cargo meeting today and the conclusion was:

* The semantics here seem good. They match `make` for the build phase and other CLI tools for everything else.
* Let's add a CI check of some form that `println!` and `eprintln!` aren't part of the codebase
* We should confirm what happens when `make` detects stdout isn't there and there are also active jobs elsewhere.

ehuss commented:
I added a test to check for print macros (and added dbg! for fun).  It is not bullet-proof, but should be good enough.

I checked `make`, and it does the same as Cargo when it tries to print to stdout and it is closed (fails the build, prints `Waiting for unfinished jobs`, etc.).

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 5a096da9d7035c8d35c4050742835317933662b6 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 5a096da9d7035c8d35c4050742835317933662b6 -->
<!-- homu: {"type":"Approved","sha":"5a096da9d7035c8d35c4050742835317933662b6","approver":"alexcrichton"} -->

bors commented:
:hourglass: Testing commit 5a096da9d7035c8d35c4050742835317933662b6 with merge cb7cff15c8b0a78fccdc5d797485490dc3bd6c85...
<!-- homu: {"type":"BuildStarted","head_sha":"5a096da9d7035c8d35c4050742835317933662b6","merge_sha":"cb7cff15c8b0a78fccdc5d797485490dc3bd6c85"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/473adcf1-6656-4c7e-9337-c393e7e0431c/_build/results?buildId=29171)
Approved by: alexcrichton
Pushing cb7cff15c8b0a78fccdc5d797485490dc3bd6c85 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"alexcrichton","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/473adcf1-6656-4c7e-9337-c393e7e0431c/_build/results?buildId=29171"},"merge_sha":"cb7cff15c8b0a78fccdc5d797485490dc3bd6c85"} -->
